sql/schemachanger: Validate ALTER TABLE .. TYPE in the DSC

spilchen opened the pull request:
When altering the data type of a column, some changes only require validating the existing data. For example, when changing from `CHAR(20)` to `CHAR(10)`, we don't need to rewrite the data; we only need to ensure all existing data fits within `CHAR(10)`. This change implements that validation logic in the declarative schema changer.

The validation logic uses a temporary check constraint. This constraint is added transiently so that it never transitions to PUBLIC and is automatically removed. The check constraint casts the column to the new type and then back to the old type. If the value matches its original value after these casts, the row is deemed valid.

To add a schema changer test, I disabled the
`enable_experimental_alter_column_type_general` setting for validation. It remains enabled for column type alterations requiring a rewrite and for validation-only changes in the legacy schema changer.

I updated the dependency rules to allow the check constraint to be added transiently. I think this is fine even for older releases because I'm not introducing new elements or attributes for the check constraint; I'm just allowing a state transition to transient.

When validation occurs, it will look something like this:

```
demo@127.0.0.1:26257/demoapp/movr> ALTER TABLE t1 ALTER COLUMN c1 SET
DATA TYPE CHAR(3);
ERROR: validation of CHECK "CAST(CAST(c1 AS CHAR(3)) AS CHAR(20)) = c1"
failed on row: c1='123456789', rowid=990187907647504385
SQLSTATE: 23514
```

Epic: CRDB-25314
Closes: #127516
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/127823)
<!-- Reviewable:end -->

annrpom reviewed (commented):
i think the changes overall make sense

although we don't enable the dsc in explicit txns yet, i wanted to make note/ask a question about a schema change like this in an explicit txn:

```
-- some stuff i ran before repro:
set use_declarative_schema_changer = unsafe_always; 
create table t(i char(30));
insert into t values (123456789);

demo@127.0.0.1:26257/demoapp/movr> begin;                                    
BEGIN

Time: 0ms total (execution 0ms / network 0ms)

demo@127.0.0.1:26257/demoapp/movr  OPEN> ALTER TABLE t ALTER COLUMN i SET    
                                      -> DATA TYPE CHAR(3);                  
ALTER TABLE

Time: 9ms total (execution 8ms / network 0ms)

demo@127.0.0.1:26257/demoapp/movr  OPEN> insert into t values (1);           
INSERT 0 1

Time: 4ms total (execution 4ms / network 0ms)

demo@127.0.0.1:26257/demoapp/movr  OPEN> commit;                             
ERROR: transaction committed but schema change aborted with error: (23514): validation of CHECK "CAST(CAST(i AS CHAR(3)) AS CHAR(30)) = i" failed on row: i='123456789', rowid=990583138430386177
SQLSTATE: XXA00
HINT: Some of the non-DDL statements may have committed successfully, but some of the DDL statement(s) failed.
Manual inspection may be required to determine the actual state of the database.
--
See: https://go.crdb.dev/issue-v/42061/v24.2
```

dsc + deprules is a bit of a black box for me, but: are we able to be more strict on the when the constraint check occurs? it's a bit of a ux surprise right now

spilchen reviewed (commented):
You raise a good point. I modeled the usage of a check constraint on dropping column NULLability. It produces similar output at commit:
```
demo@127.0.0.1:26257/demoapp/movr> set use_declarative_schema_changer = unsafe_always;
                                -> create table t(i char(30));
                                -> insert into t values (NULL);
SET

Time: 1ms total (execution 0ms / network 0ms)

CREATE TABLE

Time: 4ms total (execution 4ms / network 0ms)

INSERT 0 1

Time: 4ms total (execution 4ms / network 0ms)

demo@127.0.0.1:26257/demoapp/movr> begin;
BEGIN

Time: 0ms total (execution 0ms / network 0ms)

demo@127.0.0.1:26257/demoapp/movr  OPEN> alter table t alter column i set not null;
ALTER TABLE

Time: 5ms total (execution 5ms / network 0ms)

demo@127.0.0.1:26257/demoapp/movr  OPEN> commit;
ERROR: transaction committed but schema change aborted with error: (23502): validation of column "i" NOT NULL failed on row: i=NULL, rowid=990764130680799233
SQLSTATE: XXA00
HINT: Some of the non-DDL statements may have committed successfully, but some of the DDL statement(s) failed.
Manual inspection may be required to determine the actual state of the database.
--
See: https://go.crdb.dev/issue-v/42061/v24.2
```

Maybe more of a @fqazi question, but should we look at doing the check constraint at statement time rather than at commit?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/127823#-:-O37hs642u_5AwwAxEWG:be4tei6)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @wenyihu6)

*[a discussion](https://reviewable.io/reviews/cockroachdb/cockroach/127823#gh-1695452790:-O37hq4v6OmGeo7-99Om:bb6un1j) (no related file):*
<details><summary><i>Previously, annrpom (annie pompa) wrote…</i></summary><blockquote>

*(Reviewable was [unable](https://github.com/Reviewable/Reviewable/issues/14) to map [this GitHub inline comment thread](https://github.com/cockroachdb/cockroach/pull/127823#discussion_r1695452790) to the right spot — sorry!)*

nit: *we skip (perhaps?)

> So if skip when running the legacy schema changer.

So we skip when running the legacy schema changer.
</blockquote></details>

Done

___


<!-- Sent from Reviewable.io -->

annrpom commented:
ooo actually i think `set autocommit_before_ddl = on;` would also do the trick

annrpom reviewed (approved):
i took a second pass at this and have one comment:

not any new behavior introduced by this PR, but while we are at it: do we want to put some notice when changing the type of a column to a type that differs from its child/parent:
```
-- demo of a notice we have if we just try to create a FK between char(30) and char(31)
demo@127.0.0.1:26257/demoapp/movr> CREATE TABLE t1 (                                        
                                ->     col1 CHAR(30) UNIQUE                                 
                                -> );                                                       
CREATE TABLE

Time: 6ms total (execution 5ms / network 0ms)

demo@127.0.0.1:26257/demoapp/movr> CREATE TABLE t2 (                                        
                                ->     t1_col1 CHAR(31),                                    
                                ->     col2 INT,                                            
                                ->     FOREIGN KEY (t1_col1) REFERENCES t1(col1)            
                                -> );                                                       
NOTICE: type of foreign key column "t1_col1" (char) is not identical to referenced column "t1"."col1" (char)
CREATE TABLE

Time: 40ms total (execution 40ms / network 0ms)
```
```
-- this is what currently happens:
demo@127.0.0.1:26257/demoapp/movr> CREATE TABLE t1 (                                        
                                ->     col1 CHAR(7) UNIQUE                                  
                                -> );                                                       
CREATE TABLE

Time: 8ms total (execution 6ms / network 2ms)

demo@127.0.0.1:26257/demoapp/movr> CREATE TABLE t2 (                                        
                                ->     t1_col1 CHAR(7),                                     
                                ->     col2 INT,                                            
                                ->     FOREIGN KEY (t1_col1) REFERENCES t1(col1)            
                                -> );                                                       
CREATE TABLE

Time: 51ms total (execution 51ms / network 0ms)

-- a notice here would be nice
demo@127.0.0.1:26257/demoapp/movr> alter table t2 alter column t1_col1 set data type        
                                -> char(5);                                                 
ALTER TABLE

Time: 85ms total (execution 85ms / network 0ms)
```

spilchen reviewed (commented):
Good catch with the notice. I wasn't aware that we generated that. I've updated the commit to now detect this. I also changed the output slightly to make it more understandable. In the example you provided, we referred to the types as the same:
```
NOTICE: type of foreign key column "t1_col1" (char) is not identical to referenced column "t1"."col1" (char)
```

I used a different function to get the type name, so now it will print out the size of the `char`, making the difference more obvious. The new notice will look like this:
```
NOTICE: type of foreign key column "t1_col1" (CHAR(31)) is not identical to referenced column "t1"."col1" (CHAR(30))
```

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/127823#-:-O3D5j7K-oz8Jb9KTyr9:b-bmd8pe)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @wenyihu6)



<!-- Sent from Reviewable.io -->

fqazi reviewed (changes requested):
Just a few minor comments.

> Reviewed 16 of 18 files at r1, 1 of 1 files at r2, 4 of 4 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/127823#-:-O3E71JHEyqrbdAhDgsj:byqjw8p)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @spilchen and @wenyihu6)

___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_alter_column_type.go` line 197 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/127823#-O3Dw7qi3SmGtxqpLFEB:-O3Dw7qi3SmGtxqpLFEC:b-tk4est) ([raw file](https://github.com/cockroachdb/cockroach/blob/cb8cf094edd4fa78316e9d3080e635bfea7d109e/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_alter_column_type.go#L197)):*
> ```Go
> 	// The constraint requires a backing index to use, which we will use the
> 	// primary index.
> 	chain := getPrimaryIndexChain(b, newColType.TableID)
> ```

Nit: We might want to turn the logic of picking the index below as a helper. I think add column does the same thing.
___
*[`pkg/sql/schemachanger/scplan/internal/rules/current/dep_alter_column_type.go` line 43 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/127823#-O3Dy_1x5tEzyLMnW72T:-O3Dy_1yCEmPWZGhibtb:b-qptrsp) ([raw file](https://github.com/cockroachdb/cockroach/blob/cb8cf094edd4fa78316e9d3080e635bfea7d109e/pkg/sql/schemachanger/scplan/internal/rules/current/dep_alter_column_type.go#L43)):*
> ```Go
> 	registerDepRule(
> 		"column type is changed to public after doing validation of a transient check constraint",
> 		scgraph.Precedence,
> ```

I think this one needs to be strict. Needs to happen in the SameStagePrecedence to prevent stuff from sneaking in after validation.
___
*[`pkg/sql/schemachanger/scplan/internal/rules/current/dep_alter_column_type.go` line 52 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/127823#-O3Dy3E5A1mkcfKAaGEk:-O3Dy3E5A1mkcfKAaGEl:bn9u5sd) ([raw file](https://github.com/cockroachdb/cockroach/blob/cb8cf094edd4fa78316e9d3080e635bfea7d109e/pkg/sql/schemachanger/scplan/internal/rules/current/dep_alter_column_type.go#L52)):*
> ```Go
> 				to.El.AttrEqVar(screl.ColumnID, colID),
> 				from.ReferencedColumnIDsContains(colID),
> 				JoinOnDescID(from, to, "table-id"),
> ```

Nit: Maybe move this one up.  think these get evaluated in order.
___
*[`pkg/sql/schemachanger/testdata/end_to_end/alter_table_alter_column_type_validation/alter_table_alter_column_type_validation.explain` line 59 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/127823#-O3DyW6QEcjH86XJbrDy:-O3DyW6QEcjH86XJbrDz:b-83sjoh) ([raw file](https://github.com/cockroachdb/cockroach/blob/cb8cf094edd4fa78316e9d3080e635bfea7d109e/pkg/sql/schemachanger/testdata/end_to_end/alter_table_alter_column_type_validation/alter_table_alter_column_type_validation.explain#L59)):*
> ```
>            │    └── ABSENT              → PUBLIC           ColumnType:{DescID: 104 (t), ColumnFamilyID: 0 (primary), ColumnID: 2 (j), TypeName: "CHAR(5)"}
>            ├── 1 element transitioning toward TRANSIENT_ABSENT
>            │    └── TRANSIENT_VALIDATED → TRANSIENT_ABSENT CheckConstraint:{DescID: 104 (t), IndexID: 1 (t_pkey), ConstraintID: 2, ReferencedColumnIDs: [2]}
> ```

Nice!
___
*[`pkg/ccl/changefeedccl/cdcevent/event_test.go` line 444 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/127823#-O3E6zbYDiz-cUaX3c98:-O3E6zbYDiz-cUaX3c99:bdsv95q) ([raw file](https://github.com/cockroachdb/cockroach/blob/cb8cf094edd4fa78316e9d3080e635bfea7d109e/pkg/ccl/changefeedccl/cdcevent/event_test.go#L444)):*
> ```Go
> 	// Use the legacy schema changer for now, as the declarative schema changer
> 	// cannot do backfill yet. This will be fixed in issue #127014.
> 	sqlDB.Exec(t, `SET use_declarative_schema_changer = off`)
> ```

Shouldn't we just fallback?


<!-- Sent from Reviewable.io -->

spilchen committed:
sql/schemachanger: Validate ALTER TABLE .. TYPE in the DSC

When altering the data type of a column, some changes only require
validating the existing data. For example, when changing from `CHAR(20)`
to `CHAR(10)`, we don't need to rewrite the data; we only need to ensure
all existing data fits within `CHAR(10)`. This change implements that
validation logic in the declarative schema changer.

The validation logic uses a temporary check constraint. This constraint
is added transiently so that it never transitions to PUBLIC and is
automatically removed. The check constraint casts the column to the new
type and then back to the old type. If the value matches its original
value after these casts, the row is deemed valid.

To add a schema changer test, I disabled the
`enable_experimental_alter_column_type_general` setting for validation.
It remains enabled for column type alterations requiring a rewrite and
for validation-only changes in the legacy schema changer.

I updated the dependency rules to allow the check constraint to be added
transiently. I think this is fine even for older releases because I'm not
introducing new elements or attributes for the check constraint; I'm
just allowing a state transition to transient.

When validation occurs, it will look something like this:

```
demo@127.0.0.1:26257/demoapp/movr> ALTER TABLE t1 ALTER COLUMN c1 SET
DATA TYPE CHAR(3);
ERROR: validation of CHECK "CAST(CAST(c1 AS CHAR(3)) AS CHAR(20)) = c1"
failed on row: c1='123456789', rowid=990187907647504385
SQLSTATE: 23514
```

Epic: CRDB-25314
Closes: #127516
Release note: None

spilchen reviewed (commented):
Thanks for looking @fqazi. I addressed your comments.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/127823#-:-O3IQ7F66cYXbqlU8p75:bb3nc0e)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @fqazi and @wenyihu6)

___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_alter_column_type.go` line 197 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/127823#-O3Dw7qi3SmGtxqpLFEB:-O3HwoaD0hYFXHpRHmgb:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/cb8cf094edd4fa78316e9d3080e635bfea7d109e/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_alter_column_type.go#L197)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

Nit: We might want to turn the logic of picking the index below as a helper. I think add column does the same thing.
</blockquote></details>

Done
___
*[`pkg/sql/schemachanger/scplan/internal/rules/current/dep_alter_column_type.go` line 43 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/127823#-O3Dy_1x5tEzyLMnW72T:-O3HwsQ69oPZvgNc_uoB:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cb8cf094edd4fa78316e9d3080e635bfea7d109e/pkg/sql/schemachanger/scplan/internal/rules/current/dep_alter_column_type.go#L43)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

I think this one needs to be strict. Needs to happen in the SameStagePrecedence to prevent stuff from sneaking in after validation.
</blockquote></details>

Done.
___
*[`pkg/sql/schemachanger/scplan/internal/rules/current/dep_alter_column_type.go` line 52 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/127823#-O3Dy3E5A1mkcfKAaGEk:-O3HwtuwA0dpwc1zBrNo:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/cb8cf094edd4fa78316e9d3080e635bfea7d109e/pkg/sql/schemachanger/scplan/internal/rules/current/dep_alter_column_type.go#L52)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

Nit: Maybe move this one up.  think these get evaluated in order.
</blockquote></details>

Done
___
*[`pkg/ccl/changefeedccl/cdcevent/event_test.go` line 444 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/127823#-O3E6zbYDiz-cUaX3c98:-O3Hx0RwCEncCi1_Royu:bap0f4d) ([raw file](https://github.com/cockroachdb/cockroach/blob/cb8cf094edd4fa78316e9d3080e635bfea7d109e/pkg/ccl/changefeedccl/cdcevent/event_test.go#L444)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

Shouldn't we just fallback?
</blockquote></details>

The test converts a STRING to a VARCHAR(16). In the legacy schema changer, this triggers a column rewrite/backfill, whereas in the DSC, it's only a validation change. The test relies on the backfill happening. Initially, I disabled the use of the DSC to reduce scope. However, I changed the test to perform an alter that will always force a backfill (e.g., STRING to DATE). This way, when the DSC supports it, the test won't need further updates.


<!-- Sent from Reviewable.io -->

fqazi reviewed (approved):
Nice :rocket: 

<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> 

> Reviewed 9 of 9 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/127823#-:-O3IQSTI4uHRH0VAS57A:b-tlapa4)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @wenyihu6)



<!-- Sent from Reviewable.io -->

spilchen commented:
tftr!

bors r=fqazi,annrpom
